mcs/class/System/Mono.AppleTls/Items.cs

```diff
@@ -98,39 +111,48 @@ static SecIdentity FindIdentity (Predicate<SecCertificate> filter)
 
 			return null;
 		}
-		
-		public static INativeObject[] QueryAsReference (SecRecord query, int max, out SecStatusCode result)
+
+		public static INativeObject [] QueryAsReference (SecRecord query, int max, out SecStatusCode result)
 		{
-			if (query == null){
+			if (query == null) {
 				result = SecStatusCode.Param;
 				return null;
 			}
 
-			using (var copy = query.queryDict.MutableCopy ()) {
+			using (var copy = query.QueryDict.MutableCopy ()) {
 				copy.SetValue (CFBoolean.True.Handle, SecItem.ReturnRef);
 				SetLimit (copy, max);
+				return QueryAsReference (copy, out result);
+			}
+		}
 
-				IntPtr ptr;
-				result = SecItem.SecItemCopyMatching (copy.Handle, out ptr);
-				if ((result == SecStatusCode.Success) && (ptr != IntPtr.Zero)) {
-					var array = CFArray.ArrayFromHandle<INativeObject> (ptr, p => {
-						IntPtr cfType = CFType.GetTypeID (p);
-						if (cfType == SecCertificate.GetTypeID ())
-							return new SecCertificate (p, true);
-						else if (cfType == SecKey.GetTypeID ())
-							return new SecKey (p, true);
-						else if (cfType == SecIdentity.GetTypeID ())
-							return new SecIdentity (p, true);
-						else
-							throw new Exception (String.Format ("Unexpected type: 0x{0:x}", cfType));
-					});
-					return array;
-				}
+		public static INativeObject [] QueryAsReference (CFDictionary query, out SecStatusCode result)
+		{
+			if (query == null) {
+				result = SecStatusCode.Param;
 				return null;
 			}
+
+			IntPtr ptr;
+			result = SecItem.SecItemCopyMatching (query.Handle, out ptr);
+			if ((result == SecStatusCode.Success) && (ptr != IntPtr.Zero)) {
+				var array = CFArray.ArrayFromHandle<INativeObject> (ptr, p => {
```

Comment:
Not my code, I only moved it around a little bit.  I just made this private as it's only supported as called from FindIdentity().

In some situations, the `result` array may contain elements of different types.